mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala

```diff
@@ -438,8 +443,8 @@ class RowMatrix @Since("1.0.0") (
     require(m > 1, s"RowMatrix.computeCovariance called on matrix with only $m rows." +
       "  Cannot compute the covariance of a RowMatrix with <= 1 row.")
     val mean = Vectors.fromML(summary.mean)
-
-    if (rows.first().isInstanceOf[DenseVector]) {
+    // If all the rows are sparse vectors, then compute based on `computeSparseVectorCovariance`.
+    if (!isSparseMatrix) {
       computeDenseVectorCovariance(mean, n, m)
```

Comment:
OK, I was going to say, why do we treat these two cases differently? but the code already does. I think this is a better test, so I support the change.

Reply:
So you mean, we should use `computeDenseVectorCovariance` for both sparse and dense data and remove the `computeSparseVectorCovariance` method?

Reply:
No I think the change is reasonable. Just wondering why it ever treated these two paths so differently. But that is not the issue here.

Reply:
I will take a look at the two different code pathes.
As to this PR, what about just adding a param (values are "dense"/"sparse"/"auto"(just keep curent behavior)) to control which impl is used?

Reply:
Sure, I will update

Reply:
@shahidki31   This is just a idea. What is your and @srowen  thoughts on it?

There are several similar existing places in ML, like `ANOVATest`.
And I am also thinking about whether to make two impls (for dense cases using BLAS, for sparse cases using triangle-inequality) in KMeans.

Reply:
Hm, maybe, but I don't know whether the caller can meaningfully figure out how to set it. I think this change is OK - but how did you determine the correct answer for this case? In the UT?

Reply:
@srowen I have done experimentation on 10x10 dummy data with 1's and 0's. I have computed the Eigen values and Eigen vectors using both the methods. Following are the observation

1. For both sparse and dense, strong Eigen value difference is almost negligible (diff is factor of 10^-16). Huge variations come in the weak Eigen values and corresponding Eigen vectors.
2. Users are mostly interested in strong Eigen values. So, fluctuations in the weak Eigen values doesn't affect end users much.